File: app/src/main/java/net/squanchy/analytics/AnalyticsModule.java
package net.squanchy.analytics;

import android.app.Application;

import com.crashlytics.android.Crashlytics;
import com.google.firebase.analytics.FirebaseAnalytics;

import net.squanchy.proximity.ProximityProvider;
import net.squanchy.service.proximity.injection.ProximityModule;

import dagger.Module;
import dagger.Provides;

@Module(includes = ProximityModule.class)
public class AnalyticsModule {

    private final Application application;

    public AnalyticsModule(Application application) {
        this.application = application;
    }

    @Provides
    FirebaseAnalytics firebaseAnalytics() {
        return FirebaseAnalytics.getInstance(application);
    }

    @Provides
    Crashlytics crashlytics() {
        return Crashlytics.getInstance();
    }

    @Provides
    ProximityAnalytics proximityAnalytics(Application application, ProximityProvider proximityProvider) {
        return new ProximityAnalytics(application, proximityProvider);
    }

    @Provides
    Analytics analytics(FirebaseAnalytics firebaseAnalytics, Crashlytics crashlytics, ProximityAnalytics proximityAnalitics) {
        return new Analytics(firebaseAnalytics, crashlytics, proximityAnalitics);
    }
}

File: app/src/main/java/net/squanchy/analytics/AnalyticsModule.kt
package net.squanchy.analytics

import android.app.Application
import android.content.Context
import android.preference.PreferenceManager
import com.crashlytics.android.Crashlytics
import com.google.firebase.analytics.FirebaseAnalytics
import dagger.Module
import dagger.Provides
import net.squanchy.R

@Module
class AnalyticsModule {

    @Provides
    internal fun firebaseAnalytics(application: Application): FirebaseAnalytics = FirebaseAnalytics.getInstance(application)

    @Provides
    internal fun crashlytics(): Crashlytics = Crashlytics.getInstance()

    @Provides
    internal fun firstStartDetector(application: Application): FirstStartDetector {
        val preferences = application.getSharedPreferences(FIRST_START_DETECTOR_SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
        return FirstStartDetector(preferences)
    }

    @Provides
    internal fun analytics(
        application: Application,
        firebaseAnalytics: FirebaseAnalytics,
        crashlytics: Crashlytics,
        firstStartDetector: FirstStartDetector
    ): Analytics {
        if (analyticsDisabledByUser(application)) {
            return DisabledAnalytics()
        }

        return EnabledAnalytics(firebaseAnalytics, crashlytics, firstStartDetector)
    }

    private fun analyticsDisabledByUser(application: Application): Boolean {
        val preferenceKey = application.getString(R.string.disable_analytics_key)
        return PreferenceManager.getDefaultSharedPreferences(application)
            .getBoolean(preferenceKey, false)
    }

    companion object {

        private const val FIRST_START_DETECTOR_SHARED_PREFERENCES_NAME = "first_start_detector"
    }
}
